memcopy.h

```diff
@@ -0,0 +1,128 @@
+
+#if (defined(__GNUC__) || defined(__clang__))
+#define MEMCPY __builtin_memcpy
+#define MEMSET __builtin_memset
+#else
+#define MEMCPY memcpy
+#define MEMSET memset
+#endif
+
+/* Memset LEN bytes in OUT with the value at OUT - 1. Return OUT + LEN. */
+static inline unsigned char *byte_memset(unsigned char *out, unsigned len) {
+    unsigned sz = sizeof(unsigned long);
+    Assert(len >= sz, "byte_memset should be called on larger chunks");
+
+    unsigned char *from = out - 1;
+    unsigned char c = *from;
+
+    /* First, deal with the case when LEN is not a multiple of SZ. */
+    MEMSET(out, c, sz);
+    unsigned rem = len % sz;
+    len /= sz;
+    out += rem;
+    from += rem;
+
+    while (len--) {
+        /* When sz is a constant, the compiler replaces __builtin_memset with an
+           inline version that does not incur a function call overhead. */
+        MEMSET(out, c, sz);
+        out += sz;
+    }
+
+    return out;
+}
+
+/* Byte by byte semantics: copy LEN bytes from OUT + DIST and write them to OUT. Return OUT + LEN. */
+static inline unsigned char *chunk_memcpy(unsigned char *out, unsigned char *from, unsigned len) {
+    unsigned sz = sizeof(unsigned long);
+    Assert(len >= sz, "chunk_memcpy should be called on larger chunks");
+
+    /* Copy a few bytes to make sure the loop below has a multiple of SZ bytes to be copied. */
+    unsigned long chunk;
+    MEMCPY(&chunk, from, sz);
+    MEMCPY(out, &chunk, sz);
+
+    unsigned rem = len % sz;
+    len /= sz;
+    out += rem;
+    from += rem;
+
+    while (len--) {
+        MEMCPY(&chunk, from, sz);
+        MEMCPY(out, &chunk, sz);
+        from += sz;
+        out += sz;
+    }
+
+    return out;
+}
+
+/* Copy DIST bytes from OUT - DIST into OUT + DIST * k, for 0 <= k < LEN/DIST. Return OUT + LEN. */
+static inline unsigned char *chunk_memset(unsigned char *out, unsigned char *from, unsigned dist, unsigned len) {
+    if (dist > len)
+        return chunk_memcpy(out, from, len);
+
+    Assert(len >= sizeof(unsigned long), "chunk_memset should be called on larger chunks");
+
+    /* Double up the size of the memset pattern until reaching the largest pattern of size less than SZ. */
+    unsigned long chunk = 0;
+    unsigned sz = sizeof(chunk);
+    while (dist < len && dist < sz) {
+        MEMCPY(&chunk, from, sz);
+        MEMCPY(out, &chunk, sz);
+
+        out += dist;
+        len -= dist;
+        dist += dist;
+
+        /* Make sure the next memcpy has at least SZ bytes to be copied.  */
+        if (len >= sz)
+          continue;
+
+        /* Finish up byte by byte when there are not enough bytes left. */
+        while (len--)
+            *out++ = *from++;
+        return out;
+    }
+
+    return chunk_memcpy(out, from, len);
+}
+
+/* Byte by byte semantics: copy LEN bytes from FROM and write them to OUT. Return OUT + LEN. */
+static inline unsigned char *chunk_copy(unsigned char *out, unsigned char *from, unsigned len) {
+    int dist = (int)(from - out);
```

Comment:
[from the right account this time] I recommend keeping this logic outside of this function.  The reason is that otherwise you can end up calculating differences between pointers to different buffers, and that's not safe.